### PR TITLE
Fixed support of 0-sized arrays for linalg routines in ChainerX

### PR DESCRIPTION
**Pull request by IvanYashchuk**
All of the following functions in `chainerx.linalg` now work when given empty (zero sized) input arrays: `solve`, `inv`, `svd,` `pinv`.

Ref. #6764

### CODE REVIEW

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 1]
```diff
@@ -222,6 +222,11 @@ std::tuple<Array, Array, Array> Svd(const Array& a, bool full_matrices, bool com
     vt = Empty(vt_shape, a.dtype(), a.device());
     s = Empty(Shape{k}, a.dtype(), a.device());
 
+    if (a.shape().GetTotalSize() == 0 && compute_uv && full_matrices) {
+        u = Identity(m, a.dtype(), a.device());
+        vt = Identity(n, a.dtype(), a.device());
+    }
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: How about doing this inside the kernels?

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 2]
```diff
@@ -310,6 +315,13 @@ Array PseudoInverse(const Array& a, float rcond) {
 
     std::tie(u, s, vt) = Svd(a, /*full_matrices=*/false, /*compute_uv=*/true);
 
+    // Computing the maximum along zero-sized axis is not supported
+    // therefore return earlier
+    if (a.shape().GetTotalSize() == 0) {
+        // Copy instead of new empty array is used so that backward does not raise errors
+        return Copy(a.Transpose());
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: Do you need to transpose `a`?

File: chainerx_cc/chainerx/cuda/cuda_device/dot.cu [hunk 3]
```diff
@@ -140,7 +140,7 @@ public:
                     &zero,
                     internal::GetRawOffsetData(out_contiguous),
                     data_type,
-                    n,
+                    std::max(int64_t{1}, n),
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/dot.cu]: Could you just leave a comment why this change was necessary?

Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/dot.cu]: During backward computation of solve for zero-sized input, I was getting an error about incorrect argument to `GemmEx`. Docs say that this argument should not be less than one, it also says that for `lda`, `ldb`, but tests did not complain about them.

File: chainerx_cc/chainerx/routines/linalg.cc (hunk 2 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/routines/linalg.cc]: If input for `pinv` has shape MxN, the output is NxM.

**Comment by hvy** [comment]
Were we to also return early from the kernels after the identity kernel calls?

File: chainerx_cc/chainerx/cuda/cuda_device/dot.cu (hunk 3 above)
Comment (hvy) [reply chainerx_cc/chainerx/cuda/cuda_device/dot.cu]: Could you check the actual shapes of the inputs `a` and `b` to `Dot`?

Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/dot.cu]:
```
test_linalg.py::test_Solve_param_0_{b_columns=(), dtypes=('float32', 'float32'), shape=(0, 0)}[cuda:0]
GemmEx Call! N 0  M  0  K  0  ldb  1  lda  1
```
```
test_linalg.py::test_SVD_param_1_{compute_uv=True, dtype='float32', full_matrices=False, shape=(0, 3)}[cuda:0]
GemmEx Call! N 0  M  0  K  0  ldb  3  lda  3
```
where `ldb == b_cast_layout.ld`, `lda == a_cast_layout.ld`

**Comment by IvanYashchuk** [comment]
Blocked by #8007

**Comment by hvy** [comment]
It should no longer be blocked as https://github.com/chainer/chainer/pull/8020 is merged.

**Comment by hvy** [comment]
Retriggered Travis.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 4]
```diff
@@ -213,14 +213,15 @@ void SolveImpl(const Array& a, const Array& b, const Array& out) {
     auto ipiv_ptr = static_cast<int*>(internal::GetRawOffsetData(ipiv));
 
     int buffersize = 0;
-    device_internals.cusolverdn_handle().Call(GetrfBuffersize<T>, m, m, lu_ptr, m, &buffersize);
+    device_internals.cusolverdn_handle().Call(GetrfBuffersize<T>, m, m, lu_ptr, std::max(int64_t{1}, m), &buffersize);
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Not a strong opinion but how about assigning `std::max(int64_t{1}, m)` to a local variable?

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
I wrote a minor suggestion but it's not a strong opinion and it LGTM.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
There are some failures on Jenkins, could you check them?

**Comment by IvanYashchuk** [comment]
I'm sorry, Jenkins error is fixed now.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Could you fix the style error? https://jenkins.preferred.jp/job/chainer/job/chainer_pr/1962/TEST=CHAINERX_misc,label=mn1-p100/console

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Sorry but could you resolve the conflicts and make changes according to this PR? https://github.com/chainer/chainer/pull/7970

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
LGTM!
